File: aub-service/app/pricing.py
"""
VAT / pricing helpers.

Plan prices in plans.json are VAT-INCLUSIVE (the amount actually charged). For
display we break that gross figure into the net (before-VAT) price and the VAT
component, using a configurable tax rate. Displaying before-VAT does NOT change
what is charged — only how it's presented.

    net = gross / (1 + TAX_RATE)
    vat = gross - net

TAX_RATE is read from the environment (default 0.18 — Israeli VAT) and is never
hardcoded at call sites.
"""
from __future__ import annotations

import json
import os
from typing import Any, Dict, Optional


DEFAULT_TAX_RATE = 0.18


def get_tax_rate() -> float:
    """Current VAT rate as a fraction (e.g. 0.18). Falls back to the default
    on a missing/invalid value."""
    try:
        rate = float(os.getenv("TAX_RATE", str(DEFAULT_TAX_RATE)))
        return rate if rate >= 0 else DEFAULT_TAX_RATE
    except (TypeError, ValueError):
        return DEFAULT_TAX_RATE


def vat_breakdown(gross: float, discount: float = 0.0) -> Dict[str, float]:
    """Split a VAT-inclusive gross amount into net + vat, after applying an
    optional (VAT-inclusive) discount.

    Returns the original `subtotal`, the `discount` applied, and the final
    `gross` actually charged together with its `net`/`vat` components. All
    monetary values rounded to 2 decimals; `tax_rate` is echoed back for the
    client. With no discount, `subtotal == gross` so existing callers are
    unaffected.
    """
    rate = get_tax_rate()
    subtotal = max(float(gross or 0.0), 0.0)
    discount = min(max(float(discount or 0.0), 0.0), subtotal)
    final = subtotal - discount
    net = final / (1 + rate) if rate else final
    vat = final - net
    return {
        "subtotal": round(subtotal, 2),
        "discount": round(discount, 2),
        "gross": round(final, 2),
        "net": round(net, 2),
        "vat": round(vat, 2),
        "tax_rate": rate,
    }


def price_before_vat(amount: float) -> float:
    """Strip VAT from a VAT-inclusive amount: ``amount / (1 + TAX_RATE)``
    (e.g. divide by 1.18 at the default rate), rounded to 2 decimals.

    Used right before sending a charge to iCount: the PayPage adds VAT itself, so
    we hand it the before-VAT figure and the customer ends up paying the original
    VAT-inclusive price.
    """
    rate = get_tax_rate()
    amount = max(float(amount or 0.0), 0.0)
    net = amount / (1 + rate) if rate else amount
    return round(net, 2)


def load_coupons() -> Dict[str, Dict[str, Any]]:
    """Coupon catalogue from the COUPONS_JSON env var (never hardcoded).

    Expected shape (codes are matched case-insensitively):

        {"WELCOME10": {"type": "percent", "value": 10, "label": "..."},
         "SAVE50": {"type": "amount", "value": 50}}

    Returns an empty catalogue when unset or invalid (coupons simply disabled).
    """
    raw = (os.getenv("COUPONS_JSON") or "").strip()
    if not raw:
        return {}
    try:
        data = json.loads(raw)
    except (ValueError, TypeError):
        return {}
    if not isinstance(data, dict):
        return {}
    return {
        str(code).strip().upper(): meta
        for code, meta in data.items()
        if isinstance(meta, dict)
    }


def _compute_discount(coupon: Dict[str, Any], gross: float) -> float:
    """Discount amount (VAT-inclusive) for a coupon against a gross total."""
    ctype = str(coupon.get("type", "")).lower()
    try:
        value = float(coupon.get("value", 0))
    except (TypeError, ValueError):
        return 0.0
    if value <= 0:
        return 0.0
    if ctype == "percent":
        discount = gross * (value / 100.0)
    elif ctype in ("amount", "fixed"):
        discount = value
    else:
        return 0.0
    return round(min(max(discount, 0.0), gross), 2)


def validate_coupon(code: Optional[str], gross: float) -> Optional[Dict[str, Any]]:
    """Validate a coupon code against a gross total.

    Returns ``{code, type, value, discount, label}`` for a valid code that
    yields a positive discount, else ``None`` (unknown code / no effect).
    """
    if not code:
        return None
    coupon = load_coupons().get(str(code).strip().upper())
    if not coupon:
        return None
    discount = _compute_discount(coupon, gross)
    if discount <= 0:
        return None
    return {
        "code": str(code).strip().upper(),
        "type": str(coupon.get("type", "")).lower(),
        "value": float(coupon.get("value", 0)),
        "discount": discount,
        "label": coupon.get("label"),
    }

File: frontend/src/utils/pricing.ts
/**
 * VAT-aware price helpers.
 *
 * Plan prices in config (e.g. "₪99") are VAT-INCLUSIVE - that is the amount
 * actually charged. For display we show the BEFORE-VAT (net) figure as the
 * headline plus the VAT component and the final total:
 *
 *     net = gross / (1 + TAX_RATE)
 *     vat = gross - net
 *
 * TAX_RATE is read from the environment (REACT_APP_TAX_RATE), never hardcoded
 * at call sites. Default 0.18 (Israeli VAT).
 */

export const TAX_RATE: number = (() => {
  const raw = Number(process.env.REACT_APP_TAX_RATE);
  return Number.isFinite(raw) && raw >= 0 ? raw : 0.18;
})();

/** VAT rate as a whole-number percentage, for labels like "מע״מ (18%)". */
export const TAX_RATE_PERCENT: number = Math.round(TAX_RATE * 100);

export interface PriceBreakdown {
  /** VAT-inclusive total actually charged - AFTER any coupon discount. */
  gross: number;
  /** Before-VAT (net) price of the final (discounted) total. */
  net: number;
  /** VAT amount of the final (discounted) total. */
  vat: number;
  /** Tax rate used (fraction). */
  taxRate: number;
  /** VAT-inclusive total BEFORE any discount. Defaults to `gross`. */
  subtotal?: number;
  /** VAT-inclusive coupon discount applied (0 when none). */
  discount?: number;
  /** Applied coupon code, if any. */
  couponCode?: string;
}

const round2 = (n: number): number => Math.round((n + Number.EPSILON) * 100) / 100;

/** Parse a price given as a number or a string like "₪99", "99", "₪1,299". */
export function parsePrice(price: string | number | null | undefined): number {
  if (typeof price === 'number') return Number.isFinite(price) ? price : 0;
  if (!price) return 0;
  const digits = String(price).replace(/[^\d.]/g, '');
  const n = parseFloat(digits);
  return Number.isFinite(n) ? n : 0;
}

/** Break a VAT-inclusive gross amount into net + vat. */
export function breakdownFromGross(gross: number): PriceBreakdown {
  const g = Math.max(gross || 0, 0);
  const net = TAX_RATE ? g / (1 + TAX_RATE) : g;
  return { gross: round2(g), net: round2(net), vat: round2(g - net), taxRate: TAX_RATE };
}

/** Break a price (number or "₪99" string) into net + vat. */
export function breakdownFromPrice(price: string | number | null | undefined): PriceBreakdown {
  return breakdownFromGross(parsePrice(price));
}

/**
 * Format an ILS amount. Whole numbers render without decimals (₪99); fractional
 * amounts render with exactly two (₪83.90), using Hebrew digit grouping.
 */
export function formatILS(amount: number): string {
  const safe = Number.isFinite(amount) ? amount : 0;
  const isWhole = Math.abs(safe - Math.round(safe)) < 0.005;
  const formatted = isWhole
    ? Math.round(safe).toLocaleString('he-IL')
    : safe.toLocaleString('he-IL', { minimumFractionDigits: 2, maximumFractionDigits: 2 });
  return `₪${formatted}`;
}
